src/shared/pwquality-util.c

```diff
@@ -0,0 +1,158 @@
+
+#if HAVE_PWQUALITY
+
+static void *pwquality_dl = NULL;
+
+int (*sym_pwquality_check)(pwquality_settings_t *pwq, const char *password, const char *oldpassword, const char *user, void **auxerror);
+pwquality_settings_t *(*sym_pwquality_default_settings)(void);
+void (*sym_pwquality_free_settings)(pwquality_settings_t *pwq);
+int (*sym_pwquality_generate)(pwquality_settings_t *pwq, int entropy_bits, char **password);
+int (*sym_pwquality_get_str_value)(pwquality_settings_t *pwq, int setting, const char **value);
+int (*sym_pwquality_read_config)(pwquality_settings_t *pwq, const char *cfgfile, void **auxerror);
+int (*sym_pwquality_set_int_value)(pwquality_settings_t *pwq, int setting, int value);
+const char* (*sym_pwquality_strerror)(char *buf, size_t len, int errcode, void *auxerror);
+
+int dlopen_pwquality(void) {
+        _cleanup_(dlclosep) void *dl = NULL;
+        int r;
+
+        if (pwquality_dl)
+                return 0; /* Already loaded */
+
+        dl = dlopen("libpwquality.so.1", RTLD_LAZY);
+        if (!dl)
+                return log_debug_errno(SYNTHETIC_ERRNO(EOPNOTSUPP),
+                                       "libpwquality support is not installed: %s", dlerror());
+
+        r = dlsym_many_and_warn(
+                        dl,
+                        LOG_DEBUG,
+                        &sym_pwquality_check, "pwquality_check",
+                        &sym_pwquality_default_settings, "pwquality_default_settings",
+                        &sym_pwquality_free_settings, "pwquality_free_settings",
+                        &sym_pwquality_generate, "pwquality_generate",
+                        &sym_pwquality_get_str_value, "pwquality_get_str_value",
+                        &sym_pwquality_read_config, "pwquality_read_config",
+                        &sym_pwquality_set_int_value, "pwquality_set_int_value",
+                        &sym_pwquality_strerror, "pwquality_strerror",
+                        NULL);
+        if (r < 0)
+                return r;
+
+        /* Note that we never release the reference here, because there's no real reason to, after all this
+         * was traditionally a regular shared library dependency which lives forever too. */
+        pwquality_dl = TAKE_PTR(dl);
+        return 1;
+}
+
+void pwq_maybe_disable_dictionary(pwquality_settings_t *pwq) {
+        char buf[PWQ_MAX_ERROR_MESSAGE_LEN];
+        const char *path;
+        int r;
+
+        assert(pwq);
+
+        r = sym_pwquality_get_str_value(pwq, PWQ_SETTING_DICT_PATH, &path);
+        if (r < 0) {
+                log_debug("Failed to read libpwquality dictionary path, ignoring: %s",
+                          sym_pwquality_strerror(buf, sizeof(buf), r, NULL));
+                return;
+        }
+
+        // REMOVE THIS AS SOON AS https://github.com/libpwquality/libpwquality/pull/21 IS MERGED AND RELEASED
+        if (isempty(path))
+                path = "/usr/share/cracklib/pw_dict.pwd.gz";
+
+        if (isempty(path)) {
```

Comment:
This can never be satisified, path is set two lines up.

Reply:
yay, the idea was that we'll eventually delete the three lines further up, see the `//` comment there. Then it starts making sense.

this is just moved around code, it was always like this, since it existed. i didn't want to change this in this PR too.

I mean, the reference PR did get merged and released. question is simply if all distros we care about updated things. but that's something i wanted to do in a later PR

Reply:
checking koji it appears f31,f32,f33 carry libpwquality 1.4.2 now, which is probably enough on that front. question is what is with the other distros, i.e. centos and so on.

@mrc0mmand do you have any idea about this?

do we test libpwquality support at all currently in the CI? and if so, which versions should we try to support?

Reply:
@poettering From what I know:
  * [CentOS CI] CentOS 7: **no** - I didn't add it to the dependencies, since it was used only by homed stuff and we can't enable homed, because of the ancient OpenSSL. Either way, CentOS 7 has libpwquality v1.2.3
  * [CentOS CI] Arch Linux: **yes** - v1.4.2
  * [Travis CI] Debian Testing: **yes** (only unit tests) - v1.4.2
  * [CIFuzz] Ubuntu Bionic: **no**
  * [Build test] Ubuntu Bionic: **yes** (only build) - v1.4.0
  * [FuzzBuzz] Ubuntu Bionic: **no**
  * [Semaphore CI] Debian Bullseye: **no** - libpwquality support is explicitly disabled
  * [Ubuntu CI] Ubuntu Bionic: **no** - libpwquality support is explicitly disabled

Reply:
hmm, the bionic build test probably matters enough to keep this in place for a bit longer.

thanks for all the research work. much appreciated!